router/state_broadcast.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2021 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package router
+
+import (
+	"crypto/ed25519"
+	"fmt"
+	"time"
+
+	"github.com/matrix-org/pinecone/router/events"
+	"github.com/matrix-org/pinecone/types"
+)
+
+type broadcastEntry struct {
+	Sequence types.Varu64
+	LastSeen time.Time
+}
+
+// valid returns true if the broadcast hasn't expired, or false if it has. It is
+// required for broadcasts to time out eventually, in the case that nodes leave
+// the network and return later.
+func (e *broadcastEntry) valid() bool {
+	return time.Since(e.LastSeen) < broadcastExpiryPeriod
+}
+
+// NOTE: Functions prefixed with an underscore (_) are only safe to be called
+// from the actor that owns them, in order to prevent data races.
+
+// _maintainBroadcasts sends out wakeup broadcasts to let local nodes know
+// of our presence in the network.
+func (s *state) _maintainBroadcasts() {
+	select {
+	case <-s.r.context.Done():
+		return
+	default:
+		defer s._sendBroadcastIn(wakeupBroadcastInterval)
+	}
+
+	// Clean up any broadcasts that are older than the expiry period.
+	for k, v := range s._seenBroadcasts {
+		if !v.valid() {
+			s.r.log.Println("delete old broadcast")
+			delete(s._seenBroadcasts, k)
+		}
+	}
+
+	s._sendWakeupBroadcasts()
+}
+
+func (s *state) _createBroadcastFrame() (*types.Frame, error) {
+	// Construct the broadcast packet.
+	b := frameBufferPool.Get().(*[types.MaxFrameSize]byte)
+	defer frameBufferPool.Put(b)
+	broadcast := types.WakeupBroadcast{
+		Sequence: types.Varu64(time.Now().UnixMilli()),
```

Comment:
It's not *hugely* important, but we are generally using nanosecond precision (`UnixNano`) in other places like the tree updates. Might be better to stick with the same for consistency.

Reply:
Currently the snek bootstrap frames use `UnixMilli` for sending over the wire.
This is probably since the resolution is good enough and it saves some space inside the frame as compared to `UnixNano`.
For now we'll leave them both as `UnixNano` but in the future it might be worth going through the code to see if we should consolidate on one or the other.